picard/util/pipe.py

```diff
@@ -0,0 +1,241 @@
+        if self.__is_win:
+            self.__app_version = self.__app_version.replace(".", "-")
+
+            # win32pipe.CreateNamedPipe
+            # more about the arguments: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+            self.__MAX_INSTANCES: int = 1
+            self.__BUFFER_SIZE: int = 65536
+            # timeout doesn't really matter, concurrent.futures ensures that connections are closed in declared time
+            # the value is in milliseconds
+            self.__DEFAULT_TIMEOUT: int = 300
+
+            # win32file.CreateFile
+            # more about the arguments: http://timgolden.me.uk/pywin32-docs/win32file__CreateFile_meth.html
+            self.__SHARE_MODE: int = 0
+            self.__FLAGS_AND_ATTRIBUTES: int = 0
+
+            # pywintypes.error error codes
+            # more about the error codes: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+            self.__FILE_NOT_FOUND_ERROR_CODE: int = 2
+            self.__BROKEN_PIPE_ERROR_CODE: int = 109
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+        self.permission_error_happened: bool = False
+
+        # test if pipe is listened to even if no args provided
+        if isinstance(args, list):
+            if not args:
+                args.append(self.MESSAGE_TO_IGNORE)
+        else:
+            raise ValueError("args argument MUST be a list")
+
+        if self.__is_win:
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            try:
+                self.__create_unix_pipe()
+            except FileExistsError:
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        self.__create_unix_pipe()
+                        break
+
+    def __generate_filename(self) -> str:
+        if self.__is_win:
+            prefix = "\\\\.\\pipe\\"
+        elif self.__is_mac:
+            prefix = os.path.expanduser("~/Library/Application Support/MusicBrainz/Picard/pipes/")
+        else:
+            prefix = f"{os.getenv('XDG_RUNTIME_DIR')}/"
+            # just in case the $XDG_RUNTIME_DIR is not declared, fallback dir
+            if not prefix:
+                prefix = os.path.expanduser("~/.config/MusicBrainz/Picard/pipes/")
+
+        return f"{prefix}{self.__app_name}_v{self.__app_version}_pipe_file"
+
+    def __create_unix_pipe(self) -> None:
+        try:
+            try:
+                # just to be sure that there's no broken pipe left
+                os.unlink(self.path)
+            except FileNotFoundError:
+                pass
+            try:
+                os.mkfifo(self.path)
+            # no parent dirs detected, need to create them
+            except FileNotFoundError:
+                os.makedirs(self.path)
```

Comment:
This will raise `FileExistsError` if the target already exists, it should be called with `exist_ok=True`.
Also the target isn't correct, as `self.path` is the fifo, so we need to extract the directory name.

To manipulate paths I would recommend we use [Pathlib](https://docs.python.org/3/library/pathlib.html) in this file.

Reply:
How about simply store `prefix` value as `self.__pipe_parent_dir`:
```python
        if self.__is_win:
            prefix = "\\\\.\\pipe\\"
        elif self.__is_mac:
            prefix = os.path.expanduser("~/Library/Application Support/MusicBrainz/Picard/pipes/")
        else:
            prefix = f"{os.getenv('XDG_RUNTIME_DIR')}/"
            # just in case the $XDG_RUNTIME_DIR is not declared, fallback dir
            if not prefix:
                prefix = os.path.expanduser("~/.config/MusicBrainz/Picard/pipes/")```